src/gwt/src/org/rstudio/studio/client/common/rstudioapi/AskSecretManager.java

```diff
@@ -0,0 +1,167 @@
+
+@Singleton
+public class AskSecretManager
+{
+   @Inject
+   public AskSecretManager(final RStudioAPIServerOperations server,
+                           EventBus eventBus,
+                           final GlobalDisplay globalDisplay,
+                           final Satellite satellite,
+                           final SatelliteManager satelliteManager,
+                           final DependencyManager dependencyManager)
+   {
+
+      eventBus.addHandler(AskSecretEvent.TYPE, new AskSecretEvent.Handler()
+      {
+         private boolean handleAskSecret(String targetWindow)
+         {
+            // calculate the current window name
+            String window = StringUtil.notNull(satellite.getSatelliteName());
+            
+            // handle it if the target is us
+            if (window.equals(targetWindow))
+               return true;
+            
+            // also handle if we are the main window and the specified
+            // satellite doesn't exist
+            if (!Satellite.isCurrentWindowSatellite() &&
+                !satelliteManager.satelliteWindowExists(targetWindow))
+               return true;
+            
+            // othewise don't handle
+            else
+               return false;
+         }
+         
+         @Override
+         public void onAskSecret(final AskSecretEvent e)
+         {
+            if (!handleAskSecret(e.getWindow()))
+               return;
+            
+            asksecretPending_ = true;
+
+            AskSecretDialog dialog = new AskSecretDialog(
+               e.getTitle(),
+               e.getPrompt(),
+               e.getCanRemember(),
+               e.getHasSecret(),
+               dependencyManager,
+               new ProgressOperationWithInput<AskSecretDialogResult>()
+               {
+                  @Override
+                  public void execute(final AskSecretDialogResult result,
+                                      final ProgressIndicator indicator)
+                  {
+                     asksecretPending_ = false;
+
+                     RSAEncrypt.encrypt_ServerOnly(
+                        server,
+                        result.getSecret(),
+                        new RSAEncrypt.ResponseCallback()
+                        {
+                           @Override
+                           public void onSuccess(String encryptedData)
+                           {
+                              server.askSecretCompleted(
+                                 encryptedData,
+                                 result.getRemember(),
+                                 result.getHasChanged(),
+                                 new VoidServerRequestCallback(indicator));
+                              
+                           }
+
+                           @Override
+                           public void onFailure(ServerError error)
+                           {
+                              Debug.logError(error);
```

Comment:
Since this is a waitformethod, I think we need to invoke `askSecretCompleted` even if there's an error so that the method can complete on the server side.

Reply:
For sure, the legacy password fialog also needs update, fixed.